hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/CoalescedWriteBehindQueue.java

```diff
@@ -145,9 +142,9 @@ public int drainTo(Collection<DelayedEntry> collection) {
     }
 
     @Override
-    public List<DelayedEntry> asList() {
-        Collection<DelayedEntry> values = map.values();
-        return Collections.unmodifiableList(new ArrayList<>(values));
+    public void forEach(WriteBehindQueueConsumer<DelayedEntry> consumer) {
+        consumer.size(map.size());
```

Comment:
How do we ensure that the map size doesn't change during serialisation? Thing is, we might end up with corrupted serialisation as we write the size, then the entry count changes, then we write out more or less entries than the size. Copying to list prevented this as it would stop copying once it reached the size or once the iterator was exhausted. I guess you could do the same.

For instance:
```java
    private <T> T[] copyElements(T[] a) {
        if (head < tail) {
            System.arraycopy(elements, head, a, 0, size());
        } else if (head > tail) {
            int headPortionLen = elements.length - head;
            System.arraycopy(elements, head, a, 0, headPortionLen);
            System.arraycopy(elements, 0, a, headPortionLen, tail);
        }
        return a;
    }
```
Or:
```java
    public Object[] toArray() {
        // Estimate size of array; be prepared to see more or fewer elements
        Object[] r = new Object[size()];
        Iterator<E> it = iterator();
        for (int i = 0; i < r.length; i++) {
            if (! it.hasNext()) // fewer elements than expected
                return Arrays.copyOf(r, i);
            r[i] = it.next();
        }
        return it.hasNext() ? finishToArray(r, it) : r;
    }
```